doc/source/raysgd/raysgd_pytorch.rst

```diff
@@ -3,114 +3,390 @@ RaySGD Pytorch
+  6. If there are available resources and the Trainer has fewer workers than initially specified, then it will scale up its worker pool until it reaches the initially specified ``num_workers``.
 
-Hyperparameter Optimization on Distributed Pytorch
---------------------------------------------------
+Note that we assume the Trainer itself is not on a pre-emptible node. It is currently not possible to recover from a Trainer node failure.
+
+Users can set ``checkpoint="auto"`` to always checkpoint the current model before executing a pass over the training dataset.
+
+.. code-block:: bash
+
+    trainer.train(max_retries=N, checkpoint="auto")
+
+
+Advanced: Hyperparameter Tuning
+-------------------------------
 
 ``PyTorchTrainer`` naturally integrates with Tune via the ``PyTorchTrainable`` interface. The same arguments to ``PyTorchTrainer`` should be passed into the ``tune.run(config=...)`` as shown below.
 
-.. literalinclude:: ../../../python/ray/experimental/sgd/examples/tune_example.py
+.. literalinclude:: ../../../python/ray/experimental/sgd/pytorch/examples/tune_example.py
    :language: python
    :start-after: __torch_tune_example__
 
 
-Package Reference
------------------
+Simultaneous Multi-model training
+---------------------------------
+
+In certain scenarios such as training GANs, you may want to use multiple models in the training loop. You can do this in the ``PyTorchTrainer`` by allowing the ``model_creator`` and the ``optimizer_creator`` to return multiple values. You can see the `DCGAN script <https://github.com/ray-project/ray/blob/master/python/ray/experimental/sgd/pytorch/examples/dcgan.py>`_ for an end-to-end example.
+
+.. code-block:: python
 
-.. autoclass:: ray.experimental.sgd.pytorch.PyTorchTrainer
-    :members:
+    def model_creator(config):
+        netD = Discriminator()
+        netD.apply(weights_init)
 
-    .. automethod:: __init__
+        netG = Generator()
+        netG.apply(weights_init)
+        return netD, netG
 
 
-.. autoclass:: ray.experimental.sgd.pytorch.PyTorchTrainable
-    :members:
+    def optimizer_creator(models, config):
+        net_d, net_g = models
+        discriminator_opt = optim.Adam(
+            net_d.parameters(), lr=config.get("lr", 0.01), betas=(0.5, 0.999))
+        generator_opt = optim.Adam(
+            net_g.parameters(), lr=config.get("lr", 0.01), betas=(0.5, 0.999))
+        return discriminator_opt, generator_opt
+
+
+    trainer = PyTorchTrainer(
+        model_creator,
+        data_creator,
+        optimizer_creator,
+        nn.BCELoss)
+
+.. _raysgd-custom-training:
+
+Custom Training and Validation Functions
+----------------------------------------
+
+``PyTorchTrainer`` allows you to run a custom training and validation in parallel on each worker, providing a flexibility similar to using PyTorch natively. This is done via the ``train_function`` and ``validation_function`` parameters.
```

Comment:
"custom training and validation step in ..." ? (add "step")

Reply:
```suggestion
``PyTorchTrainer`` allows you to run a custom training and validation step in parallel on each worker, providing a flexibility similar to using PyTorch natively. This is done via the ``train_function`` and ``validation_function`` parameters.
```